File: Sorting/binarySearch.js
//time: O(log n) | space: O(1)

const binarySearch = (array, target) => {

  return binarySearchHelper(array, target, 0, array.length - 1)
}

const binarySearchHelper = (array, target, left, right) => {


  while (left <= right) {
    let middle = Math.floor((left + right) / 2)
    let potenetialMatch = array[middle]
    if (potenetialMatch === target) return middle

    else if (target < potenetialMatch) right = middle - 1

    else left = middle + 1
  }
  return - 1
}

File: Patterns/binary_search.py
# Binary search (sorted array / predicate on index)
# Time: O(log n)  Space: O(1) extra (assuming feasible(mid) is O(1)).

from typing import List

def feasible(mid):
    pass

def binary_search(arr: List[int], target: int) -> int:
    left, right = 0, len(arr) - 1
    first_true_index = -1
    while left <= right:
        mid = (left + right) // 2
        if feasible(mid):
            first_true_index = mid
            right = mid - 1
        else:
            left = mid + 1

    return first_true_index
